jitstats.lua

```diff
@@ -0,0 +1,462 @@
+local bc = require("jit.bc")
+local vmdef = require("jit.vmdef")
+local jit = require("jit")
+require("table.clear")
+require("table.new")
+local jutil = jit.util or require"jit.util"
+local funcinfo, funcbc = jutil.funcinfo, jutil.funcbc
+local band = bit.band
+
+local bcnames, bcname_lookup = {}, {} 
+local traceerr_lookup = {}
+
+for i, msg in pairs(vmdef.traceerr) do
+    traceerr_lookup[msg] = i
+end
+
+local bccount = 0
+
+for i = 0, #vmdef.bcnames-1, 6 do
+    local name = string.sub(vmdef.bcnames, i, i+6):match "^%s*(.-)%s*$"
+    bcnames[bccount] = name
+    bcname_lookup[name] = bccount
+    bccount = bccount + 1
+end
+
+local traceerror_info = {
+  {name = "RECERR",  msg = "error thrown or hook called during recording"},
+  {name = "TRACEUV", msg = "trace too short"},
+  {name = "TRACEOV", msg = "trace too long"},
+  {name = "STACKOV", msg = "trace too deep"},
+  {name = "SNAPOV",  msg = "too many snapshots"},
+  {name = "BLACKL",  msg = "blacklisted"},
+  {name = "RETRY",   msg = "retry recording"},
+  {name = "NYIBC",   msg = "NYI: bytecode %d"},
+  
+  -- Recording loop ops.
+  {name = "LLEAVE",  msg = "leaving loop in root trace"},
+  {name = "LINNER",  msg = "inner loop in root trace"},
+  {name = "LUNROLL", msg = "loop unroll limit reached"},
+  
+  -- Recording calls/returns.
+  {name = "BADTYPE", msg = "bad argument type"},
+  {name = "CJITOFF", msg = "JIT compilation disabled for function"},
+  {name = "CUNROLL", msg = "call unroll limit reached"},
+  {name = "DOWNREC", msg = "down-recursion, restarting"},
+  {name = "NYIFFU",  msg = "NYI: unsupported variant of FastFunc %s"},
+  {name = "NYIRETL", msg = "NYI: return to lower frame"},
+  
+  -- Recording indexed load/store.
+  {name = "STORENN", msg = "store with nil or NaN key"},
+  {name = "NOMM",    msg = "missing metamethod"},
+  {name = "IDXLOOP", msg = "looping index lookup"},
+  {name = "NYITMIX", msg = "NYI: mixed sparse/dense table"},
+  
+  -- Recording C data operations. */
+  {name = "NOCACHE", msg = "symbol not in cache"},
+  {name = "NYICONV", msg = "NYI: unsupported C type conversion"},
+  {name = "NYICALL", msg = "NYI: unsupported C function type"},
+  
+  -- Optimizations.
+  {name = "GFAIL",   msg = "guard would always fail"},
+  {name = "PHIOV",   msg = "too many PHIs"},
+  {name = "TYPEINS", msg = "persistent type instability"},
+  
+  -- Assembler.
+  {name = "MCODEAL", msg = "failed to allocate mcode memory"},
+  {name = "MCODEOV", msg = "machine code too long"},
+  {name = "MCODELM", msg = "hit mcode limit (retrying)"},
+  {name = "SPILLOV", msg = "too many spill slots"},
+  {name = "BADRA",   msg = "inconsistent register allocation"},
+  {name = "NYIIR",   msg = "NYI: cannot assemble IR instruction %d"},
+  {name = "NYIPHI",  msg = "NYI: PHI shuffling too complex"},
+  {name = "NYICOAL", msg = "NYI: register coalescing too complex"},
+}
+
+local traceerr, traceerr_fromid = {}, table.new(#vmdef.traceerr, 0)
```

Comment:
Wouldn't this be more readable as two assignments on separate lines?

Reply:
the messages aren't too long, i'd just loose the `name`/`msg` keys, like:
```
local traceerror_info = {
  {name = "RECERR",  msg = "error thrown or hook called during recording"},
  {name = "TRACEUV", msg = "trace too short"},
  {name = "TRACEOV", msg = "trace too long"},
```

Reply:
I think you two are talking about different lines. I used named fields because I was thinking of adding more like a better display string without all the printf garbage.

Reply:
Yes, I'm not sure we're talking about the same lines.